### PR TITLE
PoC: CXSPA-7313 (PoC) Move error distinction logic to ExpressJS

### PR DESCRIPTION
**Pull request by pawelfras**
Due to upcoming changes in Angular regarding error handling in SSR/SSG (see: https://github.com/angular/angular/pull/55787) we need to adjust our solution and forward raw errors to ExpressJS and only there (with small help from `HttpErrorHandlingInterceptor`) distinguish type of error to handle it properly.

image
 
closes [CXSPA-7313](https://jira.tools.sap/browse/CXSPA-7313)

### CODE REVIEW

File: core-libs/setup/ssr/error-handling/server-error-response-handlers/server-error-response-handlers.ts [hunk 1]
```diff
@@ -4,33 +4,35 @@
...
-      !res.headersSent &&
-      err instanceof CmsPageNotFoundServerErrorResponse
-    ) {
+  (err, _req, res, _next) => {
+    if (!res.headersSent && isCmsPageNotFoundError(err)) {
       res.set('Cache-Control', 'no-store');
       res.status(404).send(documentContent);
     } else {
-      next(err);
+      res.set('Cache-Control', 'no-store');
```
Comment (Platonn) [inline core-libs/setup/ssr/error-handling/server-error-response-handlers/server-error-response-handlers.ts:25]: the `else` clause is lacking the check whether `!res.headersSent`

File: core-libs/setup/ssr/error-handling/server-error-response/propagate-server-error-response.ts [hunk 2]
```diff
@@ -5,12 +5,11 @@
...
 import { InjectionToken } from '@angular/core';
-import { CxServerErrorResponse } from './cx-server-error-response';
 
 /**
  * Function that will propagate the server error response object
  * (created during server side rendering) to the CxCommonEngine.
  */
 export const PROPAGATE_SERVER_ERROR_RESPONSE = new InjectionToken<
-  (serverResponse: CxServerErrorResponse) => void
+  (serverResponse: unknown) => void
```
Comment (Platonn) [inline core-libs/setup/ssr/error-handling/server-error-response/propagate-server-error-response.ts:14]: We no longer talk about our custom ErrorResponse obejcts. Instead, we just forward raw Errors. Therefore the `ServerResponse` vocabulary should be removed from our Epic PR

File: core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts [hunk 3]
```diff
@@ -285,7 +285,17 @@ export class OptimizedSsrEngine {
...
     const renderCallback: SsrCallbackFn = (err, html): void => {
+      /* Instead of providing another middleware for handling errors, we can handle it here.
+       * We still don't have access to the Config and OccEndpointsService so we still suffer from lack of information
+       * to check if we should map to CmsPageNotFoundServerErrorResponse
+       *
+       * IT IS RATHER BAD IDEA
+       *
+       * const mappedErr = mapErrorToServerErrorResponse(err);
+       */
+
```
Comment (Platonn) [inline core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:298]: You said `IT IS RATHER BAD IDEA` but you mentioned it here.
So I'd like to understand what pros and cons do you see for this solution?

**Comment by pawelfras** [comment]
closed in favour of https://github.com/SAP/spartacus/pull/19009

**Pull request closed without merging**